### PR TITLE
AWC First Version, collected changes

### PR DESCRIPTION
**Pull request by ndevelder**
## Description and Purpose

First PR to include generalized active wake control in ROSCO. This is meant to be a working draft PR. 

TODO List:

- [x] Add IPC saturation mode that allows IPC to violate some pitch limits (like peak shaving)

## Type of change
What types of change is it?
_Select the appropriate type(s) that describe this PR_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

TODO Items General:
- [ ] Add example/test for new feature
- [ ] Update registry
- [ ] Run testing

TODO Items API Change:
- [ ] Update docs with API change
- [ ] Run update_rosco_discons.py in Test_Cases/
- [ ] Update DISCON schema

### CODE REVIEW

File: ROSCO/src/ReadSetParameters.f90 [hunk 1]
```diff
@@ -390,6 +390,14 @@ SUBROUTINE ReadControlParameterFileSub(CntrPar, zmqVar, accINFILE, accINFILE_siz
...
         CALL ParseInput(UnControllerParameters,CurLine,'PA_Damping',accINFILE(1),CntrPar%PA_Damping,ErrVar)
         CALL ReadEmptyLine(UnControllerParameters,CurLine)   
+
+        !------------ AWC input ------------
+        CALL ReadEmptyLine(UnControllerParameters,CurLine)  
+        CALL ParseInput(UnControllerParameters,CurLine,'AWC_NumModes',  accINFILE(1),CntrPar%AWC_NumModes, ErrVar)
+        CALL ParseAry(  UnControllerParameters,CurLine,'AWC_n',         CntrPar%AWC_n,     CntrPar%AWC_NumModes, accINFILE(1), ErrVar)
+        CALL ParseAry(  UnControllerParameters,CurLine,'AWC_omega',     CntrPar%AWC_omega, CntrPar%AWC_NumModes, accINFILE(1), ErrVar)
+        CALL ParseAry(  UnControllerParameters,CurLine,'AWC_amp',       CntrPar%AWC_amp,   CntrPar%AWC_NumModes, accINFILE(1), ErrVar)
+        CALL ParseAry(  UnControllerParameters,CurLine,'AWC_clockangle',CntrPar%AWC_clockangle,CntrPar%AWC_NumModes, accINFILE(1), ErrVar)
```
Comment (dzalkind) [inline ROSCO/src/ReadSetParameters.f90:400]: Do you have a DISCON file that corresponds to these added inputs?

File: ROSCO_toolbox/inputs/toolbox_schema.yaml [hunk 2]
```diff
@@ -864,6 +870,44 @@ properties:
...
+                            type: number
+                        description: AWC amplitude [rad]
+                        units: rad/s
+                        default: [0.017453292]
+                    AWC_clockangle:
+                        type: array
+                        items:
+                            type: number
+                        description: AWC clock angle [deg]
+                        units: rad/s
```
Comment (jfrederik-nrel) [inline ROSCO_toolbox/inputs/toolbox_schema.yaml:909]: In description and in example, it says the unit is degrees here, but in "units" it says rad/s. I think the former is correct and "units" should be changed to degrees here.

File: ROSCO_toolbox/inputs/toolbox_schema.yaml [hunk 3]
```diff
@@ -864,6 +870,44 @@ properties:
...
+                            type: number
+                        description: AWC frequency [rad/s]
+                        units: rad/s
+                        default: [0.314159]
+                    AWC_amp:
+                        type: array
+                        items:
+                            type: number
+                        description: AWC amplitude [rad]
+                        units: rad/s
```
Comment (jfrederik-nrel) [inline ROSCO_toolbox/inputs/toolbox_schema.yaml:902]: Units should be rad not rad/s

Comment (jfrederik-nrel) [reply ROSCO_toolbox/inputs/toolbox_schema.yaml]: And, personally, I would change the units here to degrees to make it more intuitive, but that also requires a (small) change in the controller.

File: ROSCO_toolbox/utilities.py [hunk 4]
```diff
@@ -209,6 +210,13 @@ def write_DISCON(turbine, controller, param_file='DISCON.IN', txt_filename='Cp_C
...
     file.write('{}                ! PF_Offsets     - Constant blade pitch offsets for blades 1-3 [rad]\n'.format(''.join('{:<10.8f} '.format(rosco_vt['PF_Offsets'][i]) for i in range(3))))
     file.write('\n')
+    file.write('!------- Active Wake Control -----------------------------------------------------\n')
+    file.write('{0:<12d}          ! AWC_NumModes       - Number of AWC modes\n'.format(int(rosco_vt['AWC_NumModes'])))
+    file.write('{}                ! AWC_n              - AWC azimuthal number, must be an array the size of AWC_NumModes\n'.format(write_array(rosco_vt['AWC_n'],'<4d')))
+    file.write('{}                ! AWC_omega          - AWC frequency, must be an array the size of AWC_NumModes [rad/s]\n'.format(write_array(rosco_vt['AWC_omega'],'<6.4f')))
+    file.write('{}                ! AWC_amp            - AWC amplitude, must be an array the size of AWC_NumModes [rad]\n'.format(write_array(rosco_vt['AWC_amp'],'<6.4f')))
+    file.write('{}                ! AWC_clockangle     - AWC clock angle, must be an array the size of AWC_NumModes [deg]\n'.format(write_array(rosco_vt['AWC_clockangle'],'<6.4f')))
+    file.write('\n')
```
Comment (dzalkind) [inline ROSCO_toolbox/utilities.py:219]: Can you expand on these input descriptions so it's clear to the user what they do?  I'm not sure what azimuthal number means, e.g...

File: ROSCO/src/ReadSetParameters.f90 (hunk 1 above)
Comment (ndevelder) [reply ROSCO/src/ReadSetParameters.f90:426]: This was added in a more recent commit

File: ROSCO_toolbox/utilities.py (hunk 4 above)
Comment (dzalkind) [reply ROSCO_toolbox/utilities.py:219]: @kbrown1snl, this is the best place to update input file descriptions.  Your readme might have helpful info for users to place here.

Comment (kbrown1snl) [reply ROSCO_toolbox/utilities.py:219]: @dzalkind, I transferred some of the language to utilities.py. Hopefully there's not too much text there now.

**Comment by dzalkind** [comment]
These changes have been incorporated into #230.

**Pull request closed without merging**